draft-ietf-quic-qlog-main-schema.md

```diff
@@ -592,68 +592,86 @@ JSON serialization:
-  "reference_time" value is typically the first absolute timestamp. This approach
-  uses a medium amount of characters.
-
-The first option is good for stateless loggers, the second and third for stateful
-loggers. The third option is generally preferred, since it produces smaller files
-while being easier to reason about. An example for each option can be seen in
+absolute:
+: A full absolute timestamp. This approach uses the
+  largest amount of characters. This is the default value of the "time_format"
+  field.
+
+delta:
+: A delta-encoded value, based on the previously logged value. The first event
+  in a trace typically logs the full absolute timestamp. This approach uses the
+  least amount of characters.
+
+relative:
+: A relative timestamp based on some relative time base.  The log SHOULD specify
+  a "reference_time" timestamp, either by providing it explicitly in the
+  "reference_time" field in "common_fields" ({{common-fields}}) or logging an
+  absolute time in the first event. This approach uses a medium amount of
+  characters.
+
+monotonic:
+: A relative timestamp based on an opaque monotonically nondecreasing clock.
+  The "reference_time" field MUST NOT be applied to this format, since it based
+  on an opaque clock that cannot be interpreted beyond the process that
+  instantiated it.
+
+Events in each individual trace SHOULD be logged in strictly ascending timestamp
+order (though not necessarily absolute value, for the "delta" format). Tools CAN
+sort all events on the timestamp before processing them, though are not required
+to (as this could impose a significant processing overhead). This can be a problem
+especially for multi-threaded and/or streaming loggers, who could consider using a
+separate post-processor to order qlog events in time if a tool do not provide this
+feature.
+
+There is no default base clock for formats that rely on a sharable clock. Tools
+SHOULD NOT assume the ability to derive the absolute Unix timestamp from qlog
+traces. Tools should not rely events to be be consistent across traces, even
+those generated by the same logging endpoint. For reasons of privacy, a more
+privacy-preserving reference timestamp could be chosen.
+
+
+The "absolute" format is good for stateless loggers. The other formats suit
+stateful loggers.
+
+Formats that use smaller values can produce smaller-sized file and can be easier
+to reason about during analysis. An example for each option can be seen in
 {{time-format-ex}}.
 
 ~~~~~~~~
-The absolute approach will use:
+Example "absolute" TimeFormat values:
 1500, 1505, 1522, 1588
 
-The delta approach will use:
+Example "delta" TimeFormat values:
 1500, 5, 17, 66
 
-The relative approach will:
-- set the reference_time to 1500 in "common_fields"
-- use: 0, 5, 22, 88
+Example "reference" TimeFormat values, assuming a reference_time of 1500
+in "common_fields":
+0, 5, 22, 88
+
+Example "monotonic" TimeFormat values:
+0, 5, 22, 88
```

Comment:
This enforces my confusion about why `monotonic` is needed and how it's different from `relative`... 

The way I understand it is that monotonic does not necessarily adhere to milliseconds between events? 
So say the relative values would be 
`0, 5, 22, 88`

Then the monotonic ones for the same absolute event timestamps could be
`0, 4, 17, 94`
(due to clock drift, intentional fuzzing for privacy/timing attacks, etc.)

If that's true, the examples should reflect this better.

Reply:
Relative and monotonic are both formats that use a delta encoding but that's where the similarity ends.

Per the current definition of reference format, it has to be related to something that various systems can agree on. For example, sticking January 1 2023 in all my logs and logging the duration in seconds since that date would work for a reference time. Ideally, you understand what that date is, so that you can add the time deltas in order determine their absolute timestamp. It is important to agree on a calendaring or date/time system because things like leap seconds, daylight savings, or clock skew can mean your relative-time event has negative values. This makes sorting qlogs fun.

With monotonic clocks, there is no universal calendaring or date/time. A system picks _something_ and you can  then calculate durations against that. As long a single system generates events, the durations are guaranteed to never decrement. See for example https://linux.die.net/man/3/clock_gettime where

> CLOCK_MONOTONIC
Clock that cannot be set and represents monotonic time since some unspecified starting point.

So one core difference is that reference can be negative and montonic can never.  So the reference example might want to include that.

Another core difference is that monotonic typically has no reference_time that can be communicated meaningfully. Its literally not possible to obtain it on such systems.

we could make the monotonic values different for the sake of it, if it helps avoid confusion.